File: Source/Engine/Graphics/Vulkan/VulkanSwapChain.cpp
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "VulkanSwapChain.h"
#include "VulkanTexture.h"
#include "VulkanGraphics.h"

namespace Alimer
{
    namespace
    {
        struct SwapChainSupportDetails
        {
            std::vector<VkSurfaceFormatKHR> formats;
            std::vector<VkPresentModeKHR>   presentModes;
        };

        SwapChainSupportDetails QuerySwapChainSupport(VkPhysicalDevice device, VkSurfaceKHR surface)
        {
            SwapChainSupportDetails details{};

            uint32_t count;
            vkGetPhysicalDeviceSurfaceFormatsKHR(device, surface, &count, nullptr);
            details.formats.resize(count);
            vkGetPhysicalDeviceSurfaceFormatsKHR(device, surface, &count, details.formats.data());

            // Present modes
            vkGetPhysicalDeviceSurfacePresentModesKHR(device, surface, &count, nullptr);
            details.presentModes.resize(count);
            vkGetPhysicalDeviceSurfacePresentModesKHR(device, surface, &count, details.presentModes.data());

            return details;
        }

        inline VkSurfaceFormatKHR QuerySurfaceFormat(
            const std::vector<VkSurfaceFormatKHR>& available_formats,
            VkFormat format)
        {
            if (available_formats.size() == 1 && available_formats[0].format == VK_FORMAT_UNDEFINED)
                return { VK_FORMAT_B8G8R8A8_UNORM, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR };

            for (const auto& available_format : available_formats)
            {
                if (available_format.format == format && available_format.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR)
                    return available_format;
            }

            return available_formats[0];
        }

        inline bool QueryPresentMode(const std::vector<VkPresentModeKHR>& availableModes, VkPresentModeKHR wantedMode)
        {
            VkPresentModeKHR presentMode = VK_PRESENT_MODE_FIFO_KHR;

            for (const auto& availableMode : availableModes)
            {
                if (availableMode == wantedMode)
                {
                    return true;
                }
            }

            return false;
        }

        constexpr VkPresentModeKHR ToVulkan(PresentMode mode)
        {
            switch (mode)
            {
            case PresentMode::Fifo:
                return VK_PRESENT_MODE_FIFO_KHR;
            case PresentMode::Immediate:
                return VK_PRESENT_MODE_IMMEDIATE_KHR;
            case PresentMode::Mailbox:
                return VK_PRESENT_MODE_MAILBOX_KHR;
            default:
                ALIMER_UNREACHABLE();
            }
        }

        constexpr uint32_t MinImageCountForPresentMode(VkPresentModeKHR mode)
        {
            switch (mode) {
            case VK_PRESENT_MODE_FIFO_KHR:
            case VK_PRESENT_MODE_IMMEDIATE_KHR:
                return 2;
            case VK_PRESENT_MODE_MAILBOX_KHR:
                return 3;
            default:
                ALIMER_UNREACHABLE();
            }
        }
    }

    VulkanSwapChain::VulkanSwapChain(VulkanGraphics& device_, void* windowHandle, const SwapChainCreateInfo& info)
        : SwapChain(info)
        , device(device_)
    {
        VkInstance instance = device.GetInstance();

        // Create the os-specific surface
        VkResult result = VK_SUCCESS;
#if defined(VK_USE_PLATFORM_ANDROID_KHR)
        VkAndroidSurfaceCreateInfoKHR createInfo{ VK_STRUCTURE_TYPE_ANDROID_SURFACE_CREATE_INFO_KHR };
        surfaceCreateInfo.window = (ANativeWindow*)windowHandle;
        result = vkCreateAndroidSurfaceKHR(instance, &createInfo, NULL, &surface);
#elif defined(VK_USE_PLATFORM_WIN32_KHR)
        VkWin32SurfaceCreateInfoKHR createInfo{ VK_STRUCTURE_TYPE_WIN32_SURFACE_CREATE_INFO_KHR };
        createInfo.hinstance = GetModuleHandleW(nullptr);
        createInfo.hwnd = (HWND)windowHandle;
        result = vkCreateWin32SurfaceKHR(instance, &createInfo, nullptr, &surface);
#elif defined(VK_USE_PLATFORM_IOS_MVK)
        VkIOSSurfaceCreateInfoMVK createInfo{ VK_STRUCTURE_TYPE_IOS_SURFACE_CREATE_INFO_MVK };
        createInfo.pView = windowHandle;
        result = vkCreateIOSSurfaceMVK(instance, &createInfo, nullptr, &surface);
#elif defined(VK_USE_PLATFORM_MACOS_MVK)
        VkMacOSSurfaceCreateInfoMVK createInfo { VK_STRUCTURE_TYPE_MACOS_SURFACE_CREATE_INFO_MVK };
        createInfo.pView = windowHandle;
        result = vkCreateMacOSSurfaceMVK(instance, &createInfo, nullptr, &surface);
#elif defined(VK_USE_PLATFORM_WAYLAND_KHR)
        VkWaylandSurfaceCreateInfoKHR createInfo { VK_STRUCTURE_TYPE_WAYLAND_SURFACE_CREATE_INFO_KHR };
        createInfo.display = display;
        createInfo.surface = windowHandle;
        result = vkCreateWaylandSurfaceKHR(instance, &createInfo, nullptr, &surface);
#elif defined(VK_USE_PLATFORM_XCB_KHR)
        VkXcbSurfaceCreateInfoKHR createInfo { VK_STRUCTURE_TYPE_XCB_SURFACE_CREATE_INFO_KHR };
        createInfo.connection = connection;
        createInfo.window = windowHandle;
        result = vkCreateXcbSurfaceKHR(instance, &createInfo, nullptr, &surface);
#elif defined(VK_USE_PLATFORM_HEADLESS_EXT)
        VkHeadlessSurfaceCreateInfoEXT createInfo { VK_STRUCTURE_TYPE_HEADLESS_SURFACE_CREATE_INFO_EXT };
        result = vkCreateHeadlessSurfaceEXT(instance, &createInfo, nullptr, &surface);
#endif

        if (result != VK_SUCCESS)
        {
            VK_LOG_ERROR(result, "Could not create surface");
            return;
        }

        VkBool32 supported = VK_FALSE;
        VK_CHECK(
            vkGetPhysicalDeviceSurfaceSupportKHR(device.GetPhysicalDevice(), device.GetGraphicsQueueFamily(), surface, &supported)
        );
        ALIMER_ASSERT(supported);

        ResizeBackBuffer(width, height);

        if (info.label != nullptr)
        {
            device.SetObjectName(VK_OBJECT_TYPE_SWAPCHAIN_KHR, (uint64_t)handle, info.label);
        }
    }

    VulkanSwapChain::~VulkanSwapChain()
    {
        Destroy();

        if (surface != VK_NULL_HANDLE)
        {
            vkDestroySurfaceKHR(device.GetInstance(), surface, nullptr);
            surface = VK_NULL_HANDLE;
        }
    }

    void VulkanSwapChain::Destroy()
    {
        Destroy(true);
    }

    void VulkanSwapChain::Destroy(bool destroyHandle)
    {
        VK_CHECK(vkDeviceWaitIdle(device.GetHandle()));

        for (size_t i = 0; i < imageAcquiredFences.size(); ++i)
        {
            if (imageAcquiredFenceSubmitted[i])
            {
                VkFence fence = imageAcquiredFences[i];
                if (vkGetFenceStatus(device.GetHandle(), fence) == VK_NOT_READY)
                {
                    vkWaitForFences(device.GetHandle(), 1, &fence, VK_TRUE, UINT64_MAX);
                }
            }
        }

        backBufferTextures.clear();
        imageAcquiredFenceSubmitted.clear();

        for (uint32_t i = 0; i < imageCount; i++)
        {
            vkDestroySemaphore(device.GetHandle(), imageAvailableSemaphores[i], nullptr);
            vkDestroySemaphore(device.GetHandle(), renderCompleteSemaphores[i], nullptr);
            vkDestroyFence(device.GetHandle(), imageAcquiredFences[i], nullptr);
        }

        imageAvailableSemaphores.clear();
        renderCompleteSemaphores.clear();
        imageAcquiredFences.clear();
        semaphoreIndex = 0;
        backBufferIndex = 0;
        needAcquire = true;

        if (destroyHandle &&
            handle != VK_NULL_HANDLE)
        {
            vkDestroySwapchainKHR(device.GetHandle(), handle, nullptr);
            handle = VK_NULL_HANDLE;
        }
    }

    void VulkanSwapChain::ResizeBackBuffer(uint32_t width, uint32_t height)
    {
        VkSurfaceCapabilitiesKHR caps{};
        VK_CHECK(vkGetPhysicalDeviceSurfaceCapabilitiesKHR(device.GetPhysicalDevice(), surface, &caps));

        SwapChainSupportDetails swapChainDetails = QuerySwapChainSupport(device.GetPhysicalDevice(), surface);

        VkFormat vulkanSurfaceFormat = ToVulkanFormat(colorFormat);
        VkSurfaceFormatKHR surfaceFormat = QuerySurfaceFormat(swapChainDetails.formats, vulkanSurfaceFormat);
        VkPresentModeKHR vkPresentMode = ToVulkan(presentMode);
        if (!QueryPresentMode(swapChainDetails.presentModes, vkPresentMode))
        {
            if (vkPresentMode == VK_PRESENT_MODE_MAILBOX_KHR)
            {
                if (!QueryPresentMode(swapChainDetails.presentModes, VK_PRESENT_MODE_IMMEDIATE_KHR))
                {
                    vkPresentMode = VK_PRESENT_MODE_FIFO_KHR;
                }
                else
                {
                    vkPresentMode = VK_PRESENT_MODE_IMMEDIATE_KHR;
                }
            }
            else
            {
                vkPresentMode = VK_PRESENT_MODE_FIFO_KHR;
            }
        }

        // Determine the number of images
        imageCount = MinImageCountForPresentMode(vkPresentMode);
        if ((caps.maxImageCount > 0) && (imageCount > caps.maxImageCount))
        {
            imageCount = caps.maxImageCount;
        }

        VkSwapchainKHR oldSwapchain = handle;

        VkSwapchainCreateInfoKHR info{ VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR };
        info.surface = surface;
        info.minImageCount = imageCount;
        info.imageFormat = surfaceFormat.format;
        info.imageColorSpace = surfaceFormat.colorSpace;
        info.imageExtent = {};
        info.imageArrayLayers = 1;
        info.imageUsage = VK_IMAGE_USAGE_TRANSFER_SRC_BIT | VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT;
        info.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE;
        info.queueFamilyIndexCount = 0;
        info.pQueueFamilyIndices = nullptr;
        info.preTransform = VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR;
        info.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
        info.presentMode = vkPresentMode;
        info.clipped = VK_TRUE;
        info.oldSwapchain = oldSwapchain;

        if (caps.currentExtent.width == 0xffffffff)
        {
            info.imageExtent.width = width;
            info.imageExtent.height = height;
        }
        else
        {
            info.imageExtent.width = caps.currentExtent.width;
            info.imageExtent.height = caps.currentExtent.height;
        }

        info.preTransform = caps.supportedTransforms & VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR ? VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR : caps.currentTransform;

        // Simply select the first composite alpha format available
        std::vector<VkCompositeAlphaFlagBitsKHR> compositeAlphaFlags = {
            VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR,
            VK_COMPOSITE_ALPHA_PRE_MULTIPLIED_BIT_KHR,
            VK_COMPOSITE_ALPHA_POST_MULTIPLIED_BIT_KHR,
            VK_COMPOSITE_ALPHA_INHERIT_BIT_KHR,
        };
        for (auto& compositeAlphaFlag : compositeAlphaFlags) {
            if (caps.supportedCompositeAlpha & compositeAlphaFlag) {
                info.compositeAlpha = compositeAlphaFlag;
                break;
            };
        }

        if (vkCreateSwapchainKHR(device.GetHandle(), &info, nullptr, &handle) != VK_SUCCESS)
        {
            return;
        }

        if (oldSwapchain != VK_NULL_HANDLE)
        {
            vkDestroySwapchainKHR(device.GetHandle(), oldSwapchain, nullptr);
            oldSwapchain = VK_NULL_HANDLE;
        }

        width = info.imageExtent.width;
        height = info.imageExtent.height;
        colorFormat = FromVulkanFormat(info.imageFormat);

        VK_CHECK(vkGetSwapchainImagesKHR(device.GetHandle(), handle, &imageCount, nullptr));
        std::vector<VkImage> swapchainImages(imageCount);
        VK_CHECK(vkGetSwapchainImagesKHR(device.GetHandle(), handle, &imageCount, swapchainImages.data()));
        backBufferTextures.resize(imageCount);
        imageAvailableSemaphores.resize(imageCount);
        renderCompleteSemaphores.resize(imageCount);
        imageAcquiredFences.resize(imageCount);
        imageAcquiredFenceSubmitted.resize(imageCount, false);
        semaphoreIndex = 0;
        backBufferIndex = 0;

        TextureCreateInfo textureInfo{};
        textureInfo.width = info.imageExtent.width;
        textureInfo.height = info.imageExtent.height;
        textureInfo.format = FromVulkanFormat(info.imageFormat);
        textureInfo.usage = TextureUsage::RenderTarget;

        const VkSemaphoreCreateInfo semaphoreInfo{ VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO };
        const VkFenceCreateInfo fenceInfo{ VK_STRUCTURE_TYPE_FENCE_CREATE_INFO };
        for (uint32_t i = 0; i < imageCount; i++)
        {
            VK_CHECK(vkCreateSemaphore(device.GetHandle(), &semaphoreInfo, nullptr, &imageAvailableSemaphores[i]));
            VK_CHECK(vkCreateSemaphore(device.GetHandle(), &semaphoreInfo, nullptr, &renderCompleteSemaphores[i]));
            VK_CHECK(vkCreateFence(device.GetHandle(), &fenceInfo, nullptr, &imageAcquiredFences[i]));

            // Set debug names for easy track
            if (device.DebugUtilsSupported())
            {
                device.SetObjectName(VK_OBJECT_TYPE_SEMAPHORE, (uint64_t)imageAvailableSemaphores[i], fmt::format("ImageAvailableSemaphore {}", i));
                device.SetObjectName(VK_OBJECT_TYPE_SEMAPHORE, (uint64_t)renderCompleteSemaphores[i], fmt::format("RenderCompleteSemaphore {}", i));
                device.SetObjectName(VK_OBJECT_TYPE_FENCE, (uint64_t)imageAcquiredFences[i], fmt::format("ImageAcquiredFence {}", i));
            }

            VulkanTexture* texture = new VulkanTexture(device, textureInfo, swapchainImages[i]);
            backBufferTextures[i].Reset(texture);
        }

        isMinimized = false;
    }

    void VulkanSwapChain::AfterPresent(VkResult result)
    {
        if (result == VK_ERROR_OUT_OF_DATE_KHR ||
            result == VK_SUBOPTIMAL_KHR)
        {
            Recreate();
            semaphoreIndex = imageCount - 1;
        }
        else if (result != VK_SUCCESS)
        {
            VK_LOG_ERROR(result, "Failed to present swap chain image!");
            return;
        }

        // We need to acquire next image.
        needAcquire = true;
        semaphoreIndex = (semaphoreIndex + 1) % imageCount;
    }

    void VulkanSwapChain::Recreate()
    {
        VkSurfaceCapabilitiesKHR caps;
        VK_CHECK(vkGetPhysicalDeviceSurfaceCapabilitiesKHR(device.GetPhysicalDevice(), surface, &caps));

        // Don't create swapchain when window is minimized
        if (caps.currentExtent.width == 0 ||
            caps.currentExtent.height == 0)
        {
            isMinimized = true;
            return;
        }

        Destroy(false);
        ResizeBackBuffer(caps.currentExtent.width, caps.currentExtent.height);
    }

    VkResult VulkanSwapChain::AcquireNextImage() const
    {
        if (!needAcquire) {
            return VK_SUCCESS;
        }

        uint32_t submittedImageFenceIndex = (semaphoreIndex + 1u) % static_cast<uint32_t>(imageAcquiredFenceSubmitted.size());
        if (imageAcquiredFenceSubmitted[submittedImageFenceIndex])
        {
            VkFence submittedFence = imageAcquiredFences[submittedImageFenceIndex];
            if (vkGetFenceStatus(device.GetHandle(), submittedFence) == VK_NOT_READY)
            {
                VK_CHECK(vkWaitForFences(device.GetHandle(), 1, &submittedFence, VK_TRUE, ULLONG_MAX));
            }

            VK_CHECK(vkResetFences(device.GetHandle(), 1, &submittedFence));
            imageAcquiredFenceSubmitted[submittedImageFenceIndex] = false;
        }

        VkFence imageAcquiredFence = imageAcquiredFences[semaphoreIndex];
        VkSemaphore imageAcquiredSemaphore = imageAvailableSemaphores[semaphoreIndex];

        VkResult result = vkAcquireNextImageKHR(device.GetHandle(), handle, UINT64_MAX,
            imageAcquiredSemaphore,
            imageAcquiredFence,
            &backBufferIndex);

        imageAcquiredFenceSubmitted[semaphoreIndex] = (result == VK_SUCCESS);
        if (result == VK_SUCCESS)
        {
            needAcquire = false;
        }

        return result;
    }

    TextureView* VulkanSwapChain::GetCurrentTextureView() const
    {
        if (isMinimized)
        {
            // Cannot render to minimized SwapChain
            return nullptr;
        }

        VkResult result = AcquireNextImage();
        if (result == VK_NOT_READY)
        {
            LOGE("Failed to acquire texture, did you call Present?");
            return nullptr;
        }

        return backBufferTextures[backBufferIndex]->GetDefault();
    }
}

File: src/Alimer.Graphics/Vulkan/VulkanSwapChain.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using System.Diagnostics;
using Vortice.Vulkan;
using static Vortice.Vulkan.Vulkan;

namespace Alimer.Graphics.Vulkan;

internal unsafe class VulkanSwapChain : SwapChain
{
    private readonly VulkanGraphicsDevice _device;
    private readonly VkSurfaceKHR _surface;
    private VkSwapchainKHR _handle;
    private VkSemaphore _acquireSemaphore = VkSemaphore.Null;
    private VkSemaphore _releaseSemaphore = VkSemaphore.Null;
    private VulkanTexture[]? _backbufferTextures;
    public readonly object LockObject = new();

    public VulkanSwapChain(VulkanGraphicsDevice device, ISwapChainSurface surfaceSource, in SwapChainDescription descriptor)
        : base(surfaceSource, descriptor)
    {
        _device = device;
        // Create VkSurface first.
        _surface = CreateVkSurface();

        AfterReset();

        VkSurfaceKHR CreateVkSurface()
        {
            VkSurfaceKHR surface = VkSurfaceKHR.Null;
            switch (surfaceSource.Kind)
            {
                case SwapChainSurfaceType.Win32:
                    {
                        VkWin32SurfaceCreateInfoKHR surfaceCreateInfo = new()
                        {
                            hinstance = surfaceSource.ContextHandle,
                            hwnd = surfaceSource.Handle
                        };
                        vkCreateWin32SurfaceKHR(device.Instance, &surfaceCreateInfo, null, &surface).CheckResult();
                        break;
                    }

                case SwapChainSurfaceType.Android:
                    {
                        VkAndroidSurfaceCreateInfoKHR surfaceCreateInfo = new()
                        {
                            window = surfaceSource.Handle,
                        };
                        vkCreateAndroidSurfaceKHR(device.Instance, &surfaceCreateInfo, null, &surface).CheckResult();
                        break;
                    }

                case SwapChainSurfaceType.Wayland:
                    {
                        VkWaylandSurfaceCreateInfoKHR surfaceCreateInfo = new()
                        {
                            display = surfaceSource.ContextHandle,
                            surface = surfaceSource.Handle,
                        };
                        vkCreateWaylandSurfaceKHR(device.Instance, &surfaceCreateInfo, null, &surface).CheckResult();
                        break;
                    }

                case SwapChainSurfaceType.Xcb:
                    {
                        VkXcbSurfaceCreateInfoKHR surfaceCreateInfo = new()
                        {
                            connection = surfaceSource.ContextHandle,
                            window = (uint)(nuint)surfaceSource.Handle,
                        };
                        vkCreateXcbSurfaceKHR(device.Instance, &surfaceCreateInfo, null, &surface).CheckResult();
                        break;
                    }

                case SwapChainSurfaceType.Xlib:
                    {
                        VkXlibSurfaceCreateInfoKHR surfaceCreateInfo = new()
                        {
                            dpy = surfaceSource.ContextHandle,
                            window = (nuint)surfaceSource.Handle,
                        };
                        vkCreateXlibSurfaceKHR(device.Instance, &surfaceCreateInfo, null, &surface).CheckResult();
                        break;
                    }

                case SwapChainSurfaceType.MetalLayer:
                    {
                        VkMetalSurfaceCreateInfoEXT surfaceCreateInfo = new()
                        {
                            pLayer = surfaceSource.Handle,
                        };
                        vkCreateMetalSurfaceEXT(device.Instance, &surfaceCreateInfo, null, &surface).CheckResult();
                        break;
                    }

                default:
                    throw new GraphicsException($"Vulkan: Invalid kind for surface: {surfaceSource.Kind}");
            }

            VkBool32 presentSupport = false;
            VkResult result = vkGetPhysicalDeviceSurfaceSupportKHR(
                device.PhysicalDevice,
                device.GraphicsFamily,
                surface,
                &presentSupport);

            // Present family not found, we cannot create SwapChain
            if (result != VkResult.Success || !presentSupport)
            {
                throw new GraphicsException($"Vulkan: QueueFamilyIndex {device.GraphicsFamily} doesn't support presenting surface");
            }

            return surface;
        }
    }

    /// <inheritdoc />
    public override GraphicsDevice Device => _device;
    public VkFormat VkFormat { get; }
    public VkSwapchainKHR Handle => _handle;
    public VkSemaphore AcquireSemaphore => _acquireSemaphore;
    public VkSemaphore ReleaseSemaphore => _releaseSemaphore;
    public uint AcquiredImageIndex { get; set; }
    public bool NeedAcquire { get; set; }   
    public VulkanTexture CurrentTexture => _backbufferTextures![AcquiredImageIndex];

    /// <summary>
    /// Finalizes an instance of the <see cref="VulkanSwapChain" /> class.
    /// </summary>
    ~VulkanSwapChain() => Dispose(disposing: false);

    private void AfterReset()
    {
        VkSurfaceCapabilitiesKHR caps;
        vkGetPhysicalDeviceSurfaceCapabilitiesKHR(_device.PhysicalDevice, _surface, &caps).CheckResult();

        ReadOnlySpan<VkSurfaceFormatKHR> swapchainFormats = vkGetPhysicalDeviceSurfaceFormatsKHR(_device.PhysicalDevice, _surface);
        ReadOnlySpan<VkPresentModeKHR> swapchainPresentModes = vkGetPhysicalDeviceSurfacePresentModesKHR(_device.PhysicalDevice, _surface);

        VkSurfaceFormatKHR surfaceFormat = new();
        surfaceFormat.format = ((VulkanGraphicsDevice)Device).ToVkFormat(ColorFormat);
        surfaceFormat.colorSpace = VkColorSpaceKHR.SrgbNonLinear;
        bool valid = false;

        bool allowHDR = true;
        foreach (VkSurfaceFormatKHR format in swapchainFormats)
        {
            if (!allowHDR && format.colorSpace != VkColorSpaceKHR.SrgbNonLinear)
                continue;

            if (format.format == surfaceFormat.format)
            {
                surfaceFormat = format;
                valid = true;
                break;
            }
        }

        if (!valid)
        {
            surfaceFormat.format = VkFormat.B8G8R8A8Unorm;
            surfaceFormat.colorSpace = VkColorSpaceKHR.SrgbNonLinear;
        }

        VkExtent2D swapChainExtent = default;
        if (caps.currentExtent.width != 0xFFFFFFFF &&
            caps.currentExtent.width != 0xFFFFFFFF)
        {
            swapChainExtent = caps.currentExtent;
        }
        else
        {
            swapChainExtent = new((uint)DrawableSize.Width, (uint)DrawableSize.Height);
            swapChainExtent.width = Math.Max(caps.minImageExtent.width, Math.Min(caps.maxImageExtent.width, swapChainExtent.width));
            swapChainExtent.height = Math.Max(caps.minImageExtent.height, Math.Min(caps.maxImageExtent.height, swapChainExtent.height));
        }

        VkPresentModeKHR presentMode = PresentMode.ToVk();

        {
            ReadOnlySpan<VkPresentModeKHR> kPresentModeFallbacks = stackalloc VkPresentModeKHR[3]
            {
                VkPresentModeKHR.Immediate,
                VkPresentModeKHR.Mailbox,
                VkPresentModeKHR.Fifo
            };

            // Go to the target mode.
            int modeIndex = 0;
            while (kPresentModeFallbacks[modeIndex] != presentMode)
            {
                modeIndex++;
            }

            // Find the first available fallback.
            while (!HasPresentMode(swapchainPresentModes, kPresentModeFallbacks[modeIndex]))
            {
                modeIndex++;
            }

            Debug.Assert(modeIndex < kPresentModeFallbacks.Length);
            presentMode = kPresentModeFallbacks[modeIndex];

            bool HasPresentMode(ReadOnlySpan<VkPresentModeKHR> presentModes, VkPresentModeKHR target)
            {
                foreach (VkPresentModeKHR presentMode in presentModes)
                {
                    if (presentMode == target)
                        return true;
                }

                return false;
            };

        }

        // Determine the number of images
        uint imageCount = presentMode.MinImageCountForPresentMode();
        imageCount = Math.Max(imageCount, caps.minImageCount);
        if (caps.maxImageCount != 0)
        {
            imageCount = Math.Min(imageCount, caps.maxImageCount);
        }

        VkImageUsageFlags imageUsage = VkImageUsageFlags.ColorAttachment | VkImageUsageFlags.InputAttachment | VkImageUsageFlags.TransferDst;

        VkFormatProperties formatProps;
        vkGetPhysicalDeviceFormatProperties(_device.PhysicalDevice, surfaceFormat.format, &formatProps);
        if (((formatProps.optimalTilingFeatures & VkFormatFeatureFlags.TransferSrc) != 0) ||
            ((formatProps.optimalTilingFeatures & VkFormatFeatureFlags.BlitSrc) != 0))
        {
            imageUsage |= VkImageUsageFlags.TransferSrc;
        }

        VkSwapchainCreateInfoKHR createInfo = new()
        {
            surface = _surface,
            minImageCount = imageCount,
            imageFormat = surfaceFormat.format,
            imageColorSpace = surfaceFormat.colorSpace,
            imageExtent = swapChainExtent,
            imageArrayLayers = 1,
            imageUsage = imageUsage,
            imageSharingMode = VkSharingMode.Exclusive,
            preTransform = caps.currentTransform,
            compositeAlpha = VkCompositeAlphaFlagsKHR.Opaque,
            presentMode = presentMode,
            clipped = true,
            oldSwapchain = _handle
        };

        VkResult result = vkCreateSwapchainKHR(_device.Handle, &createInfo, null, out _handle);
        if (result != VkResult.Success)
        {
            Log.Error("Vulkan: Failed to create SwapChain.");
            return;
        }

        if (createInfo.oldSwapchain.IsNotNull)
        {
            vkDestroySwapchainKHR(_device.Handle, createInfo.oldSwapchain);
        }

        NeedAcquire = true;
        ReadOnlySpan<VkImage> swapChainImages = vkGetSwapchainImagesKHR(_device.Handle, _handle);
        _backbufferTextures = new VulkanTexture[swapChainImages.Length];

        for (int i = 0; i < swapChainImages.Length; i++)
        {
            TextureDescriptor descriptor = TextureDescriptor.Texture2D(
                PixelFormat.BGRA8UnormSrgb, // createInfo.imageFormat.FromVkFormat(),
                createInfo.imageExtent.width,
                createInfo.imageExtent.height,
                usage: TextureUsage.RenderTarget,
                initialLayout: ResourceStates.RenderTarget,
                label: $"BackBuffer texture {i}"
            );

            _backbufferTextures[i] = new VulkanTexture(_device, swapChainImages[i], in descriptor);
        }

        if (_acquireSemaphore.IsNull)
        {
            vkCreateSemaphore(_device.Handle, out _acquireSemaphore).CheckResult();
        }

        if (_releaseSemaphore.IsNull)
        {
            vkCreateSemaphore(_device.Handle, out _releaseSemaphore).CheckResult();
        }
    }

    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            for (int i = 0; i < _backbufferTextures!.Length; ++i)
            {
                _backbufferTextures[i].Dispose();
            }
        }

        base.Dispose(disposing);
    }

    /// <inheitdoc />
    protected internal override void Destroy()
    {
        if (_acquireSemaphore.IsNotNull)
        {
            vkDestroySemaphore(_device.Handle, _acquireSemaphore);
        }

        if (_releaseSemaphore.IsNotNull)
        {
            vkDestroySemaphore(_device.Handle, _releaseSemaphore);
        }

        vkDestroySwapchainKHR(_device.Handle, _handle, null);
        vkDestroySurfaceKHR(_device.Instance, _surface, null);
    }

    /// <inheritdoc />
    protected override void OnLabelChanged(string newLabel)
    {
        _device.SetObjectName(VkObjectType.SurfaceKHR, _handle, newLabel);
        _device.SetObjectName(VkObjectType.SwapchainKHR, _handle, newLabel);
    }

    /// <inheritdoc />
    protected override void ResizeBackBuffer()
    {

    }

    /// <inheritdoc />
    public override Texture? GetCurrentTexture()
    {
        if (!NeedAcquire)
            return _backbufferTextures![AcquiredImageIndex];

        VkResult result = VkResult.Success;
        uint imageIndex = 0;
        lock (LockObject)
        {
            result = vkAcquireNextImageKHR(
                _device.Handle,
                Handle,
                ulong.MaxValue,
                _acquireSemaphore,
                VkFence.Null,
                out imageIndex);
        }

        if (result != VkResult.Success)
        {
            // Handle outdated error in acquire
            if (result == VkResult.SuboptimalKHR || result == VkResult.ErrorOutOfDateKHR)
            {
                Device.WaitIdle();
                //_queue.Device.UpdateSwapChain(vulkanSwapChain);
                return GetCurrentTexture();
            }
        }

        AcquiredImageIndex = imageIndex;
        NeedAcquire = false;
        return CurrentTexture;
    }
}
